refactor(ivy): remove ngBaseDef

crisbeto opened the pull request:
Removes `ngBaseDef` from the compiler and any runtime code that was still referring to it. In the cases where we'd previously generate a base def we now generate a definition for an abstract directive.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -233,21 +218,80 @@ export function extractDirectiveMetadata(
+
+  const coreModule = isCore ? undefined : '@angular/core';
+
+  // Construct the map of inputs both from the decorated fields.
```
This sentence doesn't make sense; and probably isn't really needed.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -233,21 +218,80 @@ export function extractDirectiveMetadata(
+      filterToMembersWithDecorator(decoratedElements, 'Input', coreModule), evaluator,
+      resolveInput);
+
+  // And outputs.
```
Superfluous comment.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -233,21 +218,80 @@ export function extractDirectiveMetadata(
+  const outputs = parseDecoratedFields(
+      filterToMembersWithDecorator(decoratedElements, 'Output', coreModule), evaluator,
+      resolveOutput) as{[field: string]: string};
+  // Construct the list of queries.
```
Should be an empty line before this comment to keep consistent with the other formatting in this function.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -587,6 +592,56 @@ export function extractHostBindings(
+    const hostMetaMap = evaluator.evaluate(expr);
+    if (!(hostMetaMap instanceof Map)) {
+      throw new FatalDiagnosticError(
+          ErrorCode.DECORATOR_ARG_NOT_LITERAL, expr, `Decorator host metadata must be an object`);
```
```suggestion
          ErrorCode.DECORATOR_ARG_NOT_LITERAL, expr, `Decorator host metadata must evaluate to an object`);
```

AndrewKushnir commented:
Initial set of presubmits:
- [VE Presubmit](https://test.corp.google.com/ui#id=OCL:275944252:BASE:275944457:1571698954529:3cffe091)
- [Ivy Presubmit](https://test.corp.google.com/ui#id=OCL:275944252:BASE:275948176:1571700081146:7f253a98)

AndrewKushnir commented:
Quick update after running presubmit in g3: it looks like there are some regressions, but they seem to be caused by the fact that there is a problem with inheriting inputs from base classes (annotated with `@Directive()`), which is used in this PR as a replacement for `ngBaseDef`. The problem should be resolved by PR #33131, but it's hard to tell whether it's the only problem in g3 or not. My proposal is to land PR #33131 first, rebase this PR and run g3 presubmit again (if time permits). As an option, we can create a draft PR that would include the changes from this PR and PR #33131 to see if there are any regressions in g3. Thank you.

crisbeto commented:
Are those compile-time errors or runtime? If I remember correctly, we have a lot of inheritance tests for the runtime which are passing.

AndrewKushnir commented:
@crisbeto yes, the errors happen at runtime, so I guess I was incorrect assuming that #33131 will fix those. Let me collect more information and get back with specific use-cases that regressed. Thank you.

kara commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -19,42 +19,79 @@ import {AnalysisOutput, CompileResult, DecoratorHandler, DetectResult, HandlerPr
-        metadata: decorator,
-      };
+      const angularField = this.reflector.getMembersOfClass(node).find(member => {
+        if (isNgOnChanges(member)) {
```
Let's make this all lifecycle hooks. In version 10, we plan on asking users to add a decorator to any class with an Angular lifecycle hook, so we should be consistent here.

kara commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -19,42 +19,79 @@ import {AnalysisOutput, CompileResult, DecoratorHandler, DetectResult, HandlerPr
-  detect(node: ClassDeclaration, decorators: Decorator[]|null): DetectResult<Decorator>|undefined {
+  detect(node: ClassDeclaration, decorators: Decorator[]|null):
+      DetectResult<Decorator|null>|undefined {
+    // If the class is undecorated, check if any of the fields have Angular decorators,
```
```suggestion
    // If the class is undecorated, check if any of the fields have Angular decorators or lifecycle hooks,
```

kara commented on packages/core/src/render3/jit/directive.ts:
```diff
@@ -336,10 +280,40 @@ function isViewQuery(value: any): value is Query {
+}
+
+function shouldAddAbstractDirective(type: Type<any>): boolean {
+  if (usesNgOnChanges(type)) {
```
Same here. Shouldn't it be all hooks?

crisbeto commented:
The feedback has been addressed.

kara commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -19,42 +19,88 @@ import {AnalysisOutput, CompileResult, DecoratorHandler, DetectResult, HandlerPr
+            isLifecycleHook(member, 'ngAfterViewChecked') ||
+            isLifecycleHook(member, 'ngAfterContentInit') ||
+            isLifecycleHook(member, 'ngAfterContentChecked')) {
+          return true;
```
Move hooks into an array like we did for FIELD_DECORATORS?

```typescript
if (LIFECYCLE_HOOKS.some(hook => isLifecycleHook(member, hook)) {
```

kara reviewed (approved):
LGTM, with one stylistic nit

crisbeto commented:
Addressed the last comment.

kara reviewed (approved):
LGTM

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -19,42 +19,87 @@ import {AnalysisOutput, CompileResult, DecoratorHandler, DetectResult, HandlerPr
+    // hooks, and if they do, label the class as an abstract directive.
+    if (!decorators) {
+      const angularField = this.reflector.getMembersOfClass(node).find(member => {
+        if (LIFECYCLE_HOOKS.some(hookName => isLifecycleHook(member, hookName))) {
```
```suggestion
        if (!member.isStatic && member.kind === ClassMemberKind.Method && LIFECYCLE_HOOKS.has(member.name)) {
```

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -19,42 +19,87 @@ import {AnalysisOutput, CompileResult, DecoratorHandler, DetectResult, HandlerPr
+  analyze(node: ClassDeclaration, decorator: Decorator|null): AnalysisOutput<DirectiveHandlerData> {
+    let analysis: R3DirectiveMetadata|undefined;
+
+    if (decorator) {
```
```suggestion
    if (decorator !== null) {
```

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -108,13 +153,12 @@ export class DirectiveDecoratorHandler implements
  * the module.
  */
-export function extractDirectiveMetadata(
+export function extractDirectiveMetadataFromDecoratedClass(
```
Don't refactor this method. Instead, change `decorator` to take `Decorator|null`.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -580,9 +607,64 @@ export function extractHostBindings(
+}
+
+/** Extracts host bindings from an Angular class decorator (e.g. Directive or Component). */
+function extractHostBindingsFromDecoratorMetadata(
```
Same here. No need for such a large refactoring - the original `extractHostBindings` should be fine.

crisbeto replied:
Yeah those are unused now I just forgot to delete them the first time I pushed. They're gone now.

alxhub commented on packages/compiler/src/jit_compiler_facade.ts:
```diff
@@ -168,24 +168,6 @@ export class CompilerFacadeImpl implements CompilerFacade {
         factoryRes.factory, angularCoreEnv, sourceMapUrl, factoryRes.statements);
   }
 
-  compileBase(angularCoreEnv: CoreEnvironment, sourceMapUrl: string, facade: R3BaseMetadataFacade):
```
🍾

alxhub commented on packages/compiler/src/render3/view/api.ts:
```diff
@@ -13,41 +13,11 @@ import {ParseSourceSpan} from '../../parse_util';
+ * Information needed to compile a directive for the render3 runtime, extracted only from the class
+ * members, excluding any Angular decorator metadata.
  */
-export interface R3DirectiveMetadata {
```
No, leave this as is.

crisbeto replied:
Done.

AndrewKushnir commented:
- [VE Presubmit](http://test/OCL:276583247:BASE:276583294:1571957470486:784535e8)
- [Ivy Presubmit](http://test/OCL:276583247:BASE:276583452:1571957521894:1afe8ca9)

crisbeto commented:
The latest set of feedback has been addressed.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -19,42 +19,64 @@ import {AnalysisOutput, CompileResult, DecoratorHandler, DetectResult, HandlerPr
+  'Input', 'Output', 'ViewChild', 'ViewChildren', 'ContentChild', 'ContentChildren', 'HostBinding',
+  'HostListener'
+];
+const LIFECYCLE_HOOKS = [
```
Wrap in a `Set` for faster searching.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -109,15 +131,14 @@ export class DirectiveDecoratorHandler implements
  */
 export function extractDirectiveMetadata(
-    clazz: ClassDeclaration, decorator: Decorator, reflector: ReflectionHost,
+    clazz: ClassDeclaration, decorator: Decorator | null, reflector: ReflectionHost,
```
```suggestion
    clazz: ClassDeclaration, decorator: Decorator|null, reflector: ReflectionHost,
```

Surprised clang-format doesn't do this.

crisbeto replied:
Actually it seems to reformat it to the one with spaces when I remove the space. I'll leave it as is.

alxhub commented on packages/compiler/src/render3/view/api.ts:
```diff
@@ -85,11 +54,41 @@ export interface R3DirectiveMetadata {
    */
   outputs: {[field: string]: string};
 
+  /**
```
Undo this rearrangement :)

crisbeto commented:
Addressed more of the feedback.

AndrewKushnir commented:
@crisbeto FYI, there are some failures in CI after the latest changes...

crisbeto commented:
Blocked on https://github.com/angular/angular/pull/32987.

AndrewKushnir commented:
FYI, VE and Ivy presubmits were successful (g3 regression was resolved) at the point when I started them ([see my comment here](https://github.com/angular/angular/pull/33264#issuecomment-546133821)). I believe CI was "green" at that time as well, so I'm wondering if we can roll back recent changes in this PR (based on CR) if they are not critical, so we can unblock this PR. The remaining comments/improvements (if they are not critical) can be performed as a followup, especially given that we became blocked on #32987.

crisbeto committed:
refactor(ivy): remove ngBaseDef

Removes `ngBaseDef` from the compiler and any runtime code that was still referring to it. In the cases where we'd previously generate a base def we now generate a definition for an abstract directive.

alxhub commented:
Caretaker: these are being done preemptively. Once #32987 lands, this should be in a good state.

[Presubmit](http://test/OCL:276735680:BASE:276735745:1572030696588:240079d5)
[Ivy Presubmit](http://test/OCL:276735680:BASE:276735873:1572030752858:391515af)
